### PR TITLE
Allow registering module icons, adjust stories and tests

### PR DESCRIPTION
**Pull request by kostyalmm**
## Summary

Addresses issue #2143

## Relevant technical choices

* Add `icon` property to the Module object.
* Add `getModuleIcon` selector.
* Use the selector in `ModuleIcon` component
* Wrap relevant stories in `WithTestRegistry` and use `provideModules` methods.
* Alter `provideModules` test utility to add `icon` to the fixtures before adding them to the store.

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Comment by kostyalmm** [comment]
I understand that the issue says no change for the VRT but one change will come up due to the change [here](https://github.com/google/site-kit-wp/pull/2415/files#diff-6d33cbec4183cae98c82ae1d28df73b2c5e5c0c0987187e73446c1e33c33d2ceL289)
The `width` and `height` props were different which can't be the case now after using `size`

File: assets/js/components/ModuleIcon.js [hunk 1]
```diff
@@ -24,40 +24,28 @@ import PropTypes from 'prop-types';
...
-	adsense: AdSenseIcon,
-	analytics: AnalyticsIcon,
-	optimize: OptimizeIcon,
-	'pagespeed-insights': PageSpeedInsightsIcon,
-	'search-console': SearchConsoleIcon,
-	tagmanager: TagManagerIcon,
-};
+const { useSelect } = Data;
+
+export default function ModuleIcon( { slug, size = 33, ...props } ) {
```
Comment (felixarntz) [inline assets/js/components/ModuleIcon.js:32]: Maybe use `defaultProps` instead?

File: assets/js/googlesitekit/modules/datastore/modules.js [hunk 2]
```diff
@@ -388,6 +388,35 @@ const baseSelectors = {
...
+	 * Gets a specific module icon by slug.
+	 *
+	 * Returns a specific module icon by its slug.
+	 * Returns `null` if state is still loading or if said module doesn't exist or doesn't have an icon.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {Object} state Data store's state.
+	 * @param {string} slug  Module slug.
+	 * @return {(WPComponent|null|undefined)} A specific module's icon; `undefined` if state is still loading or if said module doesn't exist; `null` if it doesn't have an icon.
```
Comment (felixarntz) [inline assets/js/googlesitekit/modules/datastore/modules.js:401]: The explanation here is not exactly correct based on the code below: `undefined` is only returned if state is still loading. `null` is returned if the module doesn't exist or if it doesn't have an icon.

File: tests/js/utils.js [hunk 3]
```diff
@@ -204,13 +210,23 @@ export const provideModules = ( registry, extraData = [] ) => {
...
 	const moduleSlugs = coreModulesFixture.map( ( { slug } ) => slug );
 	const modules = coreModulesFixture
 		.map( ( module ) => {
+			const icon = moduleIconMap[ module.slug ];
 			if ( extraModules[ module.slug ] ) {
-				return { ...module, ...extraModules[ module.slug ] };
+				return { ...module, ...extraModules[ module.slug ], icon };
 			}
-			return { ...module };
+			return { ...module, icon };
```
Comment (felixarntz) [inline tests/js/utils.js:229]: While this technically works, it doesn't really fit here, since `receiveGetModules` has the purpose of receiving module data from the API, i.e. PHP. JS-specific data (like `icon` is) should be registered via `registerModule`. They'll also be stored separately internally in the datastore (`serverDefinitions` vs `clientDefinitions`), and we should stick to that here.

I suggest we introduce a `provideModuleRegistrations` or similar test utility function which calls `registerModule` for every one of the 6 modules by default, with the expected data, allowing for custom overrides to be optionally passed (basically similarly to how this `provideModules` works.

Tests or Storybook stories that require JS module data to be registered should then use that utility function.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Mostly LGTM, just a few points of feedback.

File: tests/js/utils.js (hunk 3 above)
Comment (kostyalmm) [reply tests/js/utils.js]: @felixarntz I have implemented the suggested test utility.

File: tests/js/utils.js [hunk 4]
```diff
@@ -190,6 +196,49 @@ export const provideUserInfo = ( registry, extraData = {} ) => {
...
+		const icon = moduleIconMap[ module.slug ];
+		let moduleData = { ...module, icon };
+		if ( extraModules[ module.slug ] ) {
+			moduleData = { ...moduleData, ...extraModules[ module.slug ] };
+		}
+		delete moduleData.slug;
+		registry.dispatch( CORE_MODULES ).registerModule( module.slug, { ...moduleData } );
+	} );
+
+	registry.dispatch( CORE_MODULES ).receiveGetModules( modules );
```
Comment (felixarntz) [inline tests/js/utils.js:239]: The implementation here is slightly off:
* It should only call `registerModule` for all six modules and not deal with populating server-side module data (`receiveGetModules`).
* Therefore it also shouldn't rely on the `coreModulesFixture`, which is only responsible for server-side module data. Basically, the main part of this function would be to replicate all six `registerModule` calls that in the production code are found within the `assets/js/modules/*/index.js` files into this single utility function, in order to allow us to use the same module registration data in Storybook that the plugin also uses in production.
* In other words, we should probably have for each one of the six modules an entry here with all arguments that are passed to the real `registerModule` call, and combine that base data with `extraData` for that module, if passed, before calling `registerModule`.

**Review by felixarntz — CHANGES_REQUESTED** [review]
LGTM, only the new `provideModuleRegistrations` utility is slightly off in what it does. It might be worth discussing this further, let me know if you have any questions/feedback.

File: tests/js/utils.js (hunk 4 above)
Comment (kostyalmm) [reply tests/js/utils.js]: I made the changes as requested. I had to call the `receiveGetModules` otherwise a number of stories would have to be overwritten. Now I am calling `provideModuleRegistrations` followed by `provideModules` which works as expected.

File: tests/js/utils.js [hunk 5]
```diff
@@ -190,6 +196,30 @@ export const provideUserInfo = ( registry, extraData = {} ) => {
...
+ */
+export const provideModuleRegistrations = ( registry, extraData = [] ) => {
+	const moduleIconMap = {
+		adsense: AdsenseIcon,
+		analytics: AnalyticsIcon,
+		optimize: OptimizeIcon,
+		'pagespeed-insights': PagespeedInsightsIcon,
+		'search-console': SearchConsoleIcon,
+		tagmanager: TagManagerIcon,
+	};
```
Comment (felixarntz) [inline tests/js/utils.js:216]: Following-up on my previous comment: While this works for this particular issue, I meant that it would make sense to provide the complete **real** module registration args here for each of the six modules, which include not only the icon, but also a couple of JS components. I'm gonna add these changes here for now.

**Review by felixarntz — CHANGES_REQUESTED** [review]
Mostly LGTM, just one Storybook-specific feedback. We'll also need to call `provideModuleRegistrations` in a few other places where now the icons are missing in Storybook, e.g. in all the module-specific settings stories.

**Review by felixarntz — APPROVED** [review]
Good to go now. My additional commits address various broken Storybook stories that were lacking some setup, some of which were already broken before in `develop`.

**Review by felixarntz — APPROVED** [review]
